### PR TITLE
[REST API] Endpoints for Settings

### PR DESCRIPTION
**Pull request by jeffstieler**
Based on : https://github.com/woothemes/woocommerce/pull/10610

In addition to the work already done, this PR:
- Moves the `/settings/<group>` and `/settings/<group>/<id>` route controllers into the same class
- Returns `settings` schema items from `/settings/<group>` and `/settings/<group>/<id>`
- Adds `/settings/<group>/batch` for bulk updating group settings
- Adds `_links` into all responses

### CODE REVIEW

**Comment by mikejolley** [comment]
@claudiosmweb

File: includes/abstracts/abstract-wc-rest-settings-api-controller.php [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+				break;
+		}
+
+		// A couple fields changed in the REST API -- we can just pass these too so old filters still work
+		$setting['desc']  = ( ! empty( $setting['description'] ) ? $setting['description'] : '' );
+		$setting['title'] = ( ! empty( $setting['label'] ) ? $setting['label'] : '' );
+
+		$value = apply_filters( 'woocommerce_admin_settings_sanitize_option', $value, $setting, $raw_value );
+		$value = apply_filters( "woocommerce_admin_settings_sanitize_option_" . $setting['id'], $value, $setting, $raw_value );
+		do_action( 'woocommerce_update_option', $setting );
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-rest-settings-api-controller.php]: Were we going to remove this hook since it was deprecated in 2.4 anyway?

**Comment by mikejolley** [comment]
For my reference before merge: https://a8c.slack.com/archives/proton/p1466175333000584

**Comment by justinshreve** [comment]
@mikejolley Re earlier:

> If I recall correctly, the idea was to have a core class which contains the array of settings core uses, and then have that a) be used to get settings for the API and b) be used to generate settings in WP Admin. So they are defined once in one location, rather than in admin and the various core setting classes

I started working on this using the products settings as my test case and got it working, but with a few issues. So here are some of my thoughts after playing with it a bit:

I think your suggestion should definitely be the eventual goal. Architecturally it makes sense to have settings defined in one place. It's clean and anything that needs to use or display them can do so from one source.

However, it's not as simple as just copying the arrays over and loading it from a new class. The tl;dr version is that "view" info is currently mixed with "logic" so it would require a bit more work to do properly.

Take `includes/admin/settings/class-wc-settings-products.php` for example...

There are extra parts in each setting that shouldn't be registered to the actual settings API. CSS class is one example of this but there are also things like `sectionend` or `sectionstart`. These are all needed for wp-admin display properly. It looks bad just spitting out the settings without it and things won't be laid out correctly.

I could include this data in the `WC_Core_Settings` class, but they are not actually related to the settings themselves, and unnecessary for the REST API. They just explain how they should be displayed in one specific UI, so they would be useless for any other context. That doesn't feel any cleaner then just registering the parts we need automatically like we are now - which is already up and working. I'd rather do it more properly...

The proper thing to do is refactor how`class-wc-settings-page.php` generates it's display so we can pull just setting data only from a `WC_Core_Settings`, but that'll take a bit more work to do right. I would rather see us focus on the other API endpoints right now -- I don't think it's as important.

So what I think we should do is:

* Keep the current method of auto registering current settings, with the understanding that it's temporary.
* Rename the `legacy` code to something that makes a little more sense (`wp_admin_page` or `wc_settings_page`) which was @jeffstieler's suggestion.
* Open an issue to rewrite/refactor `wc-settings-page.php`. It should be handle loading settings from a `WC_Core_Settings` class (the exact same data used by the REST API) and display them properly for the wp-admin context. I don't think this needs done for 2.7 considering our other priorities and the other API stuff that needs done.

I hope my write-up makes sense. If that works for you, I can finish renaming the code here, and then this branch should be up to date and working. And then I can jump to another part of the API.

**Comment by justinshreve** [comment]
Okay - I believe the Settings API is ready to go. Made some minor fixes today, and fixed the helper class code (code to register wp-admin settings to the REST API) to make it clearer.

**Pull request merged**